feature/app-state-modeling-and-reducers

letammyh opened the pull request:
State: StateLogger, GameState, AppState, GameReducer, AppReducer; AppDelegate: created store; Utilities: Loading enum

Commit:
State: StateLogger, AppState, GameState, AppReducer, GameReducer; Utilities: Loading; AppDelegate - created store

Nitewriter commented on NameGame/NameGame/State/AppState.swift:
```diff
@@ -0,0 +1,20 @@
+    init() {
+        userRecords = .notAsked
+        gameState = nil
+    }
```
Style change: add a whitespace line below the init method.

Nitewriter commented on NameGame/NameGame/State/Middleware/StateLogger.swift:
```diff
@@ -0,0 +1,16 @@
+    
+    func newState(state: AppState) {
+        print("GameState: \(String(describing: state.gameState))")
+    }
```
Style change: same thing here below the newState function.

Nitewriter commented on NameGame/NameGame/State/Model/GameState.swift:
```diff
@@ -0,0 +1,34 @@
+        self.answerQueue = answerQueue
+        self.images = Loading.notAsked
+        self.question = question
+    }
```
Style change: one more here.

Nitewriter commented on NameGame/NameGame/State/Model/Question.swift:
```diff
@@ -0,0 +1,30 @@
+        self.correctChoice = correctChoice
+        self.playerChoice = nil
+        self.status = .unanswered
+    }
```
Same thing.

Nitewriter commented on NameGame/NameGame/State/Reducer/GameReducer.swift:
```diff
@@ -0,0 +1,43 @@
+enum GameAction: Action {
+    case setImages(Loading<[UIImage]>)
+    case setStatus(GameState.Status)
+    case playerAnswersCorrectly
```
Change this action to `setScore(Int)`.

Nitewriter commented on NameGame/NameGame/State/Reducer/GameReducer.swift:
```diff
@@ -0,0 +1,43 @@
+    case setImages(Loading<[UIImage]>)
+    case setStatus(GameState.Status)
+    case playerAnswersCorrectly
+    case goToNextQuestion(Question)
```
Change this action to `setQuestion(Question)`.

Nitewriter commented on NameGame/NameGame/State/Reducer/GameReducer.swift:
```diff
@@ -0,0 +1,43 @@
+    case setStatus(GameState.Status)
+    case playerAnswersCorrectly
+    case goToNextQuestion(Question)
+    case gameTerminated
```
Let's move this action to the `AppAction` enum and change it's name to `endGame`. This way our `GameReducer` is only responsible for mutations to the current game.

Nitewriter commented on NameGame/NameGame/State/Reducer/GameReducer.swift:
```diff
@@ -0,0 +1,43 @@
+        }
+        
+        return state
+    }
```
One more new line needed here.

Nitewriter commented on NameGame/NameGame/State/Reducer/GameReducer.swift:
```diff
@@ -0,0 +1,43 @@
+final class GameReducer {
+    
+    class func handle(action: Action, state: AppState) -> AppState {
+        guard let action = action as? GameAction else {
```
Add a `let gameState = state.gameState` to the guard. That way we won't continue forward if there's no game to mutate.

Nitewriter reviewed (changes requested):
Just a few super simple change. Good work!

Commit:
AppState, StateLogger, GameState, Question, GameReducer, and AppReducer: addded whitespace line below last method
GameReducer: changed actions setScore(Int) and setQuestion(Question), remoeved gameTerminated
AppAction: added endGame

Nitewriter reviewed (approved):
LGTM! 💰
